go/libraries/doltcore/sqle/dprocedures/dolt_pull.go

```diff
@@ -15,16 +15,218 @@
+	if err != nil {
+		return noConflictsOrViolations, threeWayMerge, err
+	}
+	if !hasBranch {
+		return noConflictsOrViolations, threeWayMerge,
+			fmt.Errorf("branch %q not found on remote", pullSpec.Branch.GetPath())
+	}
+
+	var conflicts int
+	var fastForward int
+	for _, refSpec := range pullSpec.RefSpecs {
+		rsSeen := false // track invalid refSpecs
+		for _, branchRef := range branchRefs {
+			remoteTrackRef := refSpec.DestRef(branchRef)
+
+			if remoteTrackRef == nil {
+				continue
+			}
+
+			rsSeen = true
+			tmpDir, err := dbData.Rsw.TempTableFilesDir()
+			if err != nil {
+				return noConflictsOrViolations, threeWayMerge, err
+			}
+			// todo: can we pass nil for either of the channels?
+			srcDBCommit, err := actions.FetchRemoteBranch(ctx, tmpDir, pullSpec.Remote, srcDB, dbData.Ddb, branchRef, runProgFuncs, stopProgFuncs)
+			if err != nil {
+				return noConflictsOrViolations, threeWayMerge, err
+			}
+
+			// TODO: this could be replaced with a canFF check to test for error
+			err = dbData.Ddb.FastForward(ctx, remoteTrackRef, srcDBCommit)
+			if err != nil {
+				return noConflictsOrViolations, threeWayMerge, fmt.Errorf("fetch failed; %w", err)
+			}
+
+			// Only merge iff branch is current branch and there is an upstream set (pullSpec.Branch is set to nil if there is no upstream)
+			if branchRef != pullSpec.Branch {
+				continue
+			}
+
+			roots, ok := sess.GetRoots(ctx, dbName)
+			if !ok {
+				return noConflictsOrViolations, threeWayMerge, sql.ErrDatabaseNotFound.New(dbName)
+			}
+
+			mergeSpec, err := createMergeSpec(ctx, sess, dbName, apr, remoteTrackRef.String())
+			if err != nil {
+				return noConflictsOrViolations, threeWayMerge, err
+			}
+
+			msg := fmt.Sprintf("Merge branch '%s' of %s into %s", pullSpec.Branch.GetPath(), pullSpec.Remote.Url, dbData.Rsr.CWBHeadRef().GetPath())
+			ws, conflicts, fastForward, err = performMerge(ctx, sess, roots, ws, dbName, mergeSpec, apr.Contains(cli.NoCommitFlag), msg)
+			if err != nil && !errors.Is(doltdb.ErrUpToDate, err) {
+				return conflicts, fastForward, err
+			}
+
+			err = sess.SetWorkingSet(ctx, dbName, ws)
+			if err != nil {
+				return conflicts, fastForward, err
+			}
+		}
+		if !rsSeen {
+			return noConflictsOrViolations, threeWayMerge, fmt.Errorf("%w: '%s'", ref.ErrInvalidRefSpec, refSpec.GetRemRefToLocal())
+		}
+	}
+
+	tmpDir, err := dbData.Rsw.TempTableFilesDir()
+	if err != nil {
+		return noConflictsOrViolations, threeWayMerge, err
+	}
+	err = actions.FetchFollowTags(ctx, tmpDir, srcDB, dbData.Ddb, runProgFuncs, stopProgFuncs)
+	if err != nil {
+		return conflicts, fastForward, err
+	}
+
+	return conflicts, fastForward, nil
+}
+
+func pullerProgFunc(ctx context.Context, statsCh <-chan pull.Stats) {
+	for {
+		if ctx.Err() != nil {
+			return
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-statsCh:
+		default:
+		}
+	}
+}
+
+func progFunc(ctx context.Context, progChan <-chan pull.PullProgress) {
```

Comment:
Can you actually get rid of all these prog updater funcs while you're in here? You can just pass nil to the pull methods instead. These don't do anything and actually take a bunch of compute for no reason